hail/python/hail/methods/statgen.py

```diff
@@ -1697,6 +1699,323 @@ def pca(entry_expr, k=10, compute_loadings=False) -> Tuple[List[float], Table, T
+        structs = [hl.struct(row_group_number = hl.int64(idx), ndarray = block)
+                   for idx, block in enumerate(chunked_arr)]
+        ht = hl.Table.parallelize(structs)
+        ht = ht.key_by('row_group_number')
+        return ht
+
+    def hailBlanczos(A, G, k, l, q, block_size, times):
+
+        start = time.time()
+
+        h_list = []
+        G_i = G
+
+        for j in range(0, q):
+            temp = A.annotate(H_i = A.ndarray @ G_i)
+            temp = temp.annotate(G_i_intermediate = temp.ndarray.T @ temp.H_i)
+            result = temp.aggregate(hl.struct(Hi_chunks = hl.agg.collect(temp.H_i), 
+                                                G_i = hl.agg.ndarray_sum(temp.G_i_intermediate)))
+            localized_H_i = np.vstack(result.Hi_chunks)
+            h_list.append(localized_H_i)
+            G_i = result.G_i
+
+        temp = A.annotate(H_i = A.ndarray @ G_i)
+        result = temp.aggregate(hl.agg.collect(temp.H_i))
+        localized_H_i = np.vstack(result)
+        h_list.append(localized_H_i)
+
+        H = np.hstack(h_list) 
+
+        end = time.time()
+        q_loop_time = end - start
+
+        start = time.time()
+        
+        # perform QR decomposition on unblocked version of H
+        Q, R = np.linalg.qr(H)
+
+        end = time.time()
+        process_Q_time = end - start
+
+        A = A.annotate(part_size = A.ndarray.shape[0])
+        A = A.annotate(rows_preceeding = hl.int32(hl.scan.sum(A.part_size)))
+        A = A.annotate_globals(Qt = Q.T)
+        T = A.annotate(ndarray = A.Qt[:, A.rows_preceeding:A.rows_preceeding + A.part_size] @ A.ndarray)
+        arr_T = T.aggregate(hl.agg.ndarray_sum(T.ndarray))
+        
+        start = time.time()
+
+        U, S, W = np.linalg.svd(arr_T, full_matrices=False)
+
+        end = time.time()
+        svd_time = end - start
+
+        start = time.time()
+
+        V = Q @ U
+
+        end = time.time()
+
+        get_V_time = end - start
+        
+        truncV = V[:,:k]
+        truncS = S[:k]
+        truncW = W[:k,:]
+            
+        if times:
+            return truncV, truncS, truncW, q_loop_time, process_Q_time, svd_time, get_V_time
+        else:
+            return truncV, truncS, truncW
+
+    U, S, V = hailBlanczos(A, G, k, l, q, block_size, report_times)
+    matrix_S = np.diag(S)
+
+    scores = V.transpose() @ matrix_S
+    eigens = S * S
+
+    hail_scores = hl.nd.array(scores)
+    hail_array_scores = hail_scores._data_array()
+    cols_and_scores = hl.zip(ht.index_globals().cols, hail_array_scores).map(lambda tup: tup[0].annotate(scores = tup[1]))
+    st = hl.Table.parallelize(cols_and_scores, key=list(mt.col_key))
+
+    us = hl.nd.array(U @ matrix_S)
+    lt = ht.select()
+    lt = lt.annotate_globals(US = us)
+    lt = lt.add_index()
+    idx = lt.key
+    lt = lt.annotate(loadings = lt.US[lt.idx,:]._data_array())
+
+    if compute_loadings:
+        return eigens, st, lt
+    else:
+        return eigens, st
```

Comment:
This is how numpy would handle a situation like this, but Hail PCA currently always returns a 3-tuple, with `None` as the third element if `compute_loadings` is false.